File: SectionMattr.Samples.Csharp/Option.cs
using System.Text.Json;

namespace SectionMattr.Samples.Csharp.Option;

public class ParsedJsonData
{
    public string title { get; set; }
}

class MyJsonParser : IMattrParser<ParsedJsonData>
{
    public MattrSection<ParsedJsonData> Parse(MattrSection<string> section, MattrSection<ParsedJsonData>[] sections)
    {
        ParsedJsonData data = JsonSerializer.Deserialize<ParsedJsonData>(section.Data);
        return NewMattr.AppendData<ParsedJsonData>(data, section);
    }
}

public class Option
{
    static void Main()
    {
        string stringContent =
    @"Content before the sections.

---

More content.

---one
title: One
---

This is section one.";

        // Set delimiter only
        Mattr<string> mattrDELIMIER = NewMattr.Sections(stringContent, NewMattr.DefaultOption().SetSectionDelimiter("---"));

        // Set parser only
        Mattr<ParsedJsonData> mattrPARSER = NewMattr.Sections(stringContent, NewMattr.DefaultOption().SetSectionParse(new MyJsonParser()));

        // Set delimiter or parser
        Mattr<ParsedJsonData> mattrOPTION1 = NewMattr.Sections(stringContent, NewMattr.DefaultOption().SetSectionDelimiter("---").SetSectionParse(new MyJsonParser()));
        // or
        Mattr<ParsedJsonData> mattrOPTION2 = NewMattr.Sections(stringContent, new MattrOption<ParsedJsonData>("---", new MyJsonParser()));
    }
}

File: SectionMattr.Samples/Option.fs
module Option

open System.Text.Json
open SectionMattr

type ParsedJsonData = { title: string }

type MyJsonParser() =
    interface IMattrParser<ParsedJsonData> with
        member _.Parse
            (
                section: MattrSection<string>,
                sections: MattrSection<ParsedJsonData> array
            ) : MattrSection<ParsedJsonData> =
            let data: ParsedJsonData = JsonSerializer.Deserialize<ParsedJsonData> section.Data
            NewMattr.AppendData data section

let stringContent =
    @"Content before the sections.

---

More content.

---one
title: One
---

This is section one."

// Set delimiter only

let mattrDELIMIER: Mattr<string> =
    NewMattr.Sections(stringContent, NewMattr.DefaultOption().SetSectionDelimiter("---"))

mattrDELIMIER |> ignore

// Set parser only

let mattrPARSER: Mattr<ParsedJsonData> =
    NewMattr.Sections(stringContent, NewMattr.DefaultOption().SetSectionParse(MyJsonParser()))

mattrPARSER |> ignore

// Set delimiter or parser

let mattrOPTION1: Mattr<ParsedJsonData> =
    NewMattr.Sections(
        stringContent,
        NewMattr
            .DefaultOption()
            .SetSectionDelimiter("---")
            .SetSectionParse(MyJsonParser())
    )

mattrOPTION1 |> ignore

// or

let mattrOPTION2: Mattr<ParsedJsonData> =
    NewMattr.Sections(
        stringContent,
        { MattrOption.SectionDelimiter = "---"
          MattrOption.SectionParse = MyJsonParser() }
    )

mattrOPTION2 |> ignore
